### DISCUSSION TITLE
Waiting and retry strategy in Selenide for locating and interacting with elements

### DISCUSSION THREAD

**Discussion by itkhanz** [selenide/selenide#2651] [Ideas]
Hi,
I would like to understand what mechanisms run when you call for example:

```
$(elementLocator)
$$(elementLocator)
```

Is it similar to driver.findElement(By locator)? If not, what checks for visibility/presence etc. are performed before returning the SelenideElement or SelenideAppiumElement? Also are there any retries performed here?

Currently, whenever I need to click or interact with some element, I am writing like:

`$(withText("Swipe")).shouldBe(visible).click(tap());`

This means I need to write shouldBe(visible) before proceeding with element actions to make sure that my element actually meets the condition in shouldBe(condition). Most of the time, we want element to be either visible or present to do any action on them. I wonder if there is a way that we can confifure this check at the Test method level or Project configuration level so we do not have to chain `.shouldBe(present)` everytime something like:

`$(withText("Swipe")).click(tap());`

where some implicit checks for visibility and interactability of found elements are perfofmed before executing the click action. I know setting this in general is not a good idea because often we want to assert for invisibility of absenece of element, but I feel this check for interactability/visibility of element should be built-in element interaction actions like click(), which should throw some exception if element is not found or interactable. Before throwing exception, it should automatically retry for the condition with set timeout.

**Comment by asolntsev** [comment]
Hi @itkhanz 
1. Methods `$(elementLocator)` and `$$(elementLocator)` do NOT webdriver at all. They only return `SelenideElement` instance that is _lazy loaded_.
2. When you call some methods on it (like `click` or `shouldBe`) then `driver.findElement(By)` will be called.
3. Usually you don't need to write `.shouldBe(visible)` before `click`. You can just call `$(withText("Swipe")).click(tap());` - method `click` will first check that the element is clickable. 
4. Recently we added a special option to skip this check, and just click the element without checks: `$("#submit").click(usingDefaultMethod().force())`
